### PR TITLE
Cloud  GUI

### PR DESCRIPTION
**Pull request by TomMc608**
This is the front end connect to the app to use the cloud image AI that is being hosed on hugging face.

### CODE REVIEW

**Comment by SamanthaCra** [comment]
@TomMc608 can you please give instructions for how we are meant to run and test this?

**Comment by TomMc608** [comment]
> @TomMc608 can you please give instructions for how we are meant to run and test this?

Added the read me

File: fullstack/src/app/cloud-details/PythonBackEnd/Dockerfile
Comment (SamanthaCra) [inline fullstack/src/app/cloud-details/PythonBackEnd/Dockerfile:1]: Can we move this to the fullstack dir and give it a .cloud extention or something? Also give the existing one a .web or .app extension

File: fullstack/src/app/cloud-details/PythonBackEnd/app.py
Comment (SamanthaCra) [inline fullstack/src/app/cloud-details/PythonBackEnd/app.py:1]: can we change the PythonBackEnd dir to python-backend for consistency please?

File: fullstack/src/app/cloud-details/ReadMe
Comment (SamanthaCra) [inline fullstack/src/app/cloud-details/ReadMe:1]: Can we change the name to README.md for consistency please.

File: fullstack/src/app/cloud-details/ReadMe [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+    
+ 2. Make sure you have docker installed on your computer.
+
+ 3. Cd in to the `\fullstack\src\app\cloud-details\PythonBackEnd>`
+
+    
+   4. In the terminal run  `docker build -t fastapi-app .` 
+    
+5. Then after its been built run `docker run -d -p 8000:8000
+    fastapi-app`
```
Comment (SamanthaCra) [inline fullstack/src/app/cloud-details/ReadMe:14]: might want to change this to a single line for consistency

File: fullstack/src/app/cloud-details/page.js [hunk 2]
```diff
@@ -0,0 +1,67 @@
+'use client';
+import React, { useState } from 'react';
+
+const CloudDetails = () => {
```
Comment (SamanthaCra) [inline fullstack/src/app/cloud-details/page.js:4]: This should be changed to export default function CloudDetails() {

File: fullstack/src/app/cloud-details/page.js [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+        <div className="mt-6 p-4 bg-white text-[--foreground] rounded-md shadow-md">
+          <h3 className="text-xl font-semibold mb-2">Prediction Result:</h3>
+          <pre className="text-sm">{result}</pre>
+        </div>
+      )}
+    </div>
+  );
+};
+
+export default CloudDetails;
```
Comment (SamanthaCra) [inline fullstack/src/app/cloud-details/page.js:147]: See earlier comment, this can be removed

File: fullstack/src/app/cloud-details/page.js
Comment (SamanthaCra) [inline fullstack/src/app/cloud-details/page.js:1]: This should be split into a page and the component should be moved into components. 
```
'use client';
import Cloud from "@/components/cloud";

export default function CloudDetails() {
  return (
    <div className="app-container">
      <Cloud />
    </div>
  );
}
```

Comment (SamanthaCra) [inline fullstack/src/app/cloud-details/page.js:1]: src/app/components/cloud.jsx
```
import React, { useState } from 'react';

const CloudDetails = () => {
const [result, setResult] = useState(''); 
const [loading, setLoading] = useState(false);
const images = [
    { id: 1, src: '/images/1.png', alt: 'Cloud Image 1' },
  ];

  const handleImageClick = async (imageSrc) => {
    try {
      setLoading(true);

      const response = await fetch(imageSrc); 
      const imageBlob = await response.blob(); 

      const formData = new FormData();
      formData.append('image', imageBlob, 'uploaded_image.jpg');

      // Send the image to Python 
      const apiResponse = await fetch('http://localhost:8000/analyze', {
        method: 'POST',
        body: formData,
      });

      const resultData = await apiResponse.json(); 

      setResult(resultData.result); 
    } catch (error) {
      console.error('Error:', error);
      setResult('Error occurred during prediction.');
    } finally {
      setLoading(false); 
    }
  };
  return (
    <div className="min-h-screen bg-[--color-background] text-[--foreground] flex flex-col items-center py-10">
      <h1 className="text-3xl font-bold mb-4">Cloud Details</h1>
      <p className="text-lg mb-6">Select an image to analyze:</p>

      <div className="flex gap-4 flex-wrap justify-center mb-6">
        {images.map((image) => (
          <img
            key={image.id}
            src={image.src}
            alt={image.alt}
            className="w-36 h-36 rounded-lg cursor-pointer transition-transform duration-300 hover:scale-110 hover:shadow-lg"
            onClick={() => handleImageClick(image.src)} 
          />
        ))}
      </div>

      {loading && <p className="text-lg text-[--secondary-foreground]">Analyzing image...</p>}

      {result && (
        <div className="mt-6 p-4 bg-white text-[--foreground] rounded-md shadow-md">
          <h3 className="text-xl font-semibold mb-2">Prediction Result:</h3>
          <pre className="text-sm">{result}</pre>
        </div>
      )}
    </div>
  );
};

export default CloudDetails;
```

File: fullstack/src/app/page.js
Comment (SamanthaCra) [inline fullstack/src/app/page.js:1]: TBC with Ben/UI update

**Review by SamanthaCra — CHANGES_REQUESTED** [review]
A few changes to think about.

File: fullstack/src/app/cloud-details/PythonBackEnd/app.py [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+    allow_origins=["http://localhost:3000"],  
+    allow_credentials=True,
+    allow_methods=["*"],
+    allow_headers=["*"], 
+)
+
+@app.post("/analyze")
+async def analyze_image(image: UploadFile = File(...)):
+    if image.content_type not in ["image/jpeg", "image/png"]:
+        raise HTTPException(status_code=400, detail="Only JPEG and PNG images are supported.")
```
Comment (SamanthaCra) [inline fullstack/src/app/cloud-details/PythonBackEnd/app.py:19]: Can this allow HEIC?

**Review by SamanthaCra — COMMENTED** [review]
image
Currently impossible to read. Upload works fine, throbber is a good addition.

**Comment by SamanthaCra** [comment]
PythonBackEnd can be moved to src/

**Review by SamanthaCra — APPROVED** [review]
Well done

**Pull request merged**